### DISCUSSION TITLE
USING LR REDUCE LR ON PLATEAU

### DISCUSSION THREAD

**Discussion by massimilianoporzio** [Project-MONAI/MONAI#1362] [Q&A]
In the file test_handler_lr_scheduler.py  :
  lr_scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=1)
  handler = LrScheduleHandler(lr_scheduler, step_transform=lambda x: val_engine.state.metrics["val_loss"])

In the code of challenge_baseline how to use it? Can anyone explain what to specify in step_transform and output_transform?
evaluator.state.metrics["val_loss"] does not work.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Could you please help share some info here? I think may need to change the expected metric key in `step_transform` according to `challenge_baseline` notebook.

Thanks.

**Comment by massimilianoporzio** [comment]
THE problem is that the trainer is a MONAI trainer not a Ignite trainer and
is training using a dictionary based dataset so I think I have to change
something but I don't know  exactly what...
As soon as possible I will share a code chunk with errors....

Il giorno lun 14 dic 2020 alle ore 09:57 Nic Ma <notifications@github.com>
ha scritto:

**Comment by Nic-Ma** [comment]
Sure, please share your code for better debug and analysis.
Thanks.

**Comment by che85** [comment]
`ReduceLROnPlateau` requires the validation loss as a parameter to the .step(..) method which is why it's falling.

https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau

This is particularly challenging if you want to use the ignite engines as it seems. When using the `SupervisedEvaluator` (https://docs.monai.io/en/latest/engines.html#monai.engines.SupervisedEvaluator), no option is giving for calculating the loss. 

You can probably find a workaround for this. The following is just quick and dirty... (monai internal code might have changed):

```
from monai.engines import SupervisedEvaluator

class SupervisedValidator(SupervisedEvaluator):

    def __init__(self, **kwargs):
        self.loss_function = kwargs.pop("loss_function")
        super(SupervisedValidator, self).__init__(**kwargs)
        self.state.metrics["val_loss"] = 2.0

    def _iteration(self, engine, batchdata):
        if batchdata is None:
            raise ValueError("Must provide batch data for current iteration.")
        batch = self.prepare_batch(batchdata, engine.state.device, engine.non_blocking)
        if len(batch) == 2:
            inputs, targets = batch
            args = tuple()
            kwargs = dict()
        else:
            inputs, targets, args, kwargs = batch

        # execute forward computation
        self.network.eval()
        with torch.no_grad():
            if self.amp:
                with torch.cuda.amp.autocast():
                    predictions = self.inferer(inputs, self.network, *args, **kwargs)
                    loss = self.loss_function(predictions, targets).mean()
            else:
                predictions = self.inferer(inputs, self.network, *args, **kwargs)
                loss = self.loss_function(predictions, targets).mean()

        self.state.metrics["val_loss"] = loss.item()
        torch.cuda.empty_cache()

        return {Keys.IMAGE: inputs, Keys.LABEL: targets, Keys.PRED: predictions, Keys.LOSS: loss.item()}
```

Probably, the post_transforms option might be another way to save the loss inside the evaluator

**Reply by Nic-Ma** [reply]
Hi @che85 ,

Thanks for your proposal and sample code!
1. To set args for the `step()` function of lr_scheduler, specify the `step_transform` in `LrScheduleHandler` and attach the handler to the `SupervisedEvaluator` engine:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/handlers/lr_schedule_handler.py#L37
2. I think maybe the `val_loss` @massimilianoporzio asked is a metric for validation? So you can just set the metrics for the SupervisedEvaluator.

Thanks.
